devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+
+    def __sub__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(sympy.Add(*[self.expr, -other.expr], evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(sympy.Add(*[self.expr, -other]), fd=self.fd)
+
+    def __rsub__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(sympy.Add(*[-self.expr, other.expr], evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(sympy.Add(*[-self.expr, other], evaluate=False),
+                                  fd=self.fd)
+
+    def __isub__(self, other):
+        if isinstance(other, Differentiable):
+            self._expr = sympy.Add(*[self.expr, -other.expr], evaluate=False)
+        else:
+            self._expr = sympy.Add(*[self.expr, -other], evaluate=False)
+        return self
+
+    def __mul__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(sympy.Mul(*[self.expr, other.expr], evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(sympy.Mul(*[self.expr, other], evaluate=False),
+                                  fd=self.fd)
+
+    __rmul__ = __mul__
+
+    def __imul__(self, other):
+        if isinstance(other, Differentiable):
+            self._expr = sympy.Mul(*[self.expr, other.expr], evaluate=False)
+        else:
+            self._expr = sympy.Mul(*[self.expr, other], evaluate=False)
+        return self
+
+    def __truediv__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(self.expr * sympy.Pow(other.expr, -1, evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(self.expr * sympy.Pow(other, -1, evaluate=False),
+                                  fd=self.fd)
+
+    def __rtruediv__(self, other):
+        if isinstance(other, Differentiable):
+            return Differentiable(other.expr * sympy.Pow(self.expr, -1, evaluate=False),
+                                  fd=self.fd)
+        else:
+            return Differentiable(other * sympy.Pow(self.expr, -1, evaluate=False),
+                                  fd=self.fd)
+
+    def __neg__(self):
+        return -1.0 * self
+
+    def __pow__(self, exponent):
+        return Differentiable(sympy.Pow(self, exponent, evaluate=False), fd=self.fd)
+
+    def __eq__(self, other):
+        if isinstance(other, Differentiable):
+            return self.expr.args == other.expr.args
+        elif isinstance(other, sympy.Expr):
+            return self.expr.args == other.args
+        else:
+            return sympy.simplify(self.expr) == other
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def func(self, *args, **kwargs):
+        kwargs.pop("evaluate", None)
+        if hasattr(self.expr, 'base'):
```

Comment:
I think we should be more precise here. I don't like using ``hasattr``, it makes it feel like an hack

Reply:
It kinda is, for some reason some `Indexed` object end up here and haven't figured out from where yet, should be here